### PR TITLE
fix errors in coordinate extractor due to duplicated coordinates

### PR DESCRIPTION
**Pull request by MoKob**
# Issue

resolves https://github.com/Project-OSRM/osrm-backend/issues/3401

## Tasklist
 - [x] add regression / cucumber cases (see docs/testing.md)
 - [x] review
 - [x] adjust for comments

## Code Review Checklist - author check these when done, reviewer verify
 - [x] Code formatted with `scripts/format.sh`
 - [x] Changes have test coverage
 - [x] New exceptions, logging, errors - are messages distinct enough to track down in the code if they get thrown in production on non-debug builds?
 - [x] The PR is one logically integrated piece of work.  If there are unrelated changes, are they at least separate commits?
 - [x] Commit messages - are they clear enough to understand the intent of the change if we have to look at them later?
 - [x] Code comments - are there comments explaining the intent?
 - [x] Relevant docs updated
 - [ ] Changelog entry if required
 - [x] Impact on the API surface
   - [x] If HTTP/libosrm.o is backward compatible features, bump the minor version
   - [x] File format changes require at minor release
   - [x] If old clients can't use the API after changes, bump the major version

If something doesn't apply, please ~~cross it out~~

## Requirements / Relations
none

### CODE REVIEW

File: features/guidance/turn-angles.feature [hunk 1]
```diff
@@ -317,30 +317,30 @@ Feature: Simple Turns
...
-                                                |
-                                                |
-                                                |
-                                                i
+                                                          h
+                                                          |
+                                                          |
+                                                          |
+                                                          |
+                                                          |      . b - - - - - - - - - - - - - - - - - - - - - - - a
```
Comment (MoKob) [inline features/guidance/turn-angles.feature]: with the fixed curve detection, this test wasn't modelled long enough. Had to change it to test the same information as before

File: features/guidance/turn-angles.feature [hunk 2]
```diff
@@ -627,12 +627,14 @@ Feature: Simple Turns
...
 
+
     #http://www.openstreetmap.org/#map=18/52.56251/13.32650
+    @todo
     Scenario: Curved Turn on Separated Directions
         Given the node map
             """
                 e                               d
-                f - - - - - - - - - - - - - - - c - - - - - - - - - - - j
+                f                               c - - - - - - - - - - - j
```
Comment (MoKob) [inline features/guidance/turn-angles.feature]: this was never in here

File: features/guidance/turn-angles.feature [hunk 3]
```diff
@@ -680,6 +682,58 @@ Feature: Simple Turns
             | j,o       | Kapweg,Kapweg,Kapweg | depart,continue uturn,arrive |
             | a,i       | Kurt,Kurt,Kurt       | depart,continue uturn,arrive |
 
+    #http://www.openstreetmap.org/#map=18/52.56251/13.32650
```
Comment (MoKob) [inline features/guidance/turn-angles.feature]: same scenario as above, but now without the additional short segment before the turn

File: src/extractor/guidance/coordinate_extractor.cpp [hunk 4]
```diff
@@ -352,40 +370,25 @@ util::Coordinate CoordinateExtractor::ExtractRepresentativeCoordinate(
          * We distinguish between turns that simply model the initial way of getting onto the
          * destination lanes and the ones that performa a larger turn.
          */
-        const double offset =
-            std::min(0.5 * considered_lanes * ASSUMED_LANE_WIDTH, 0.2 * segment_distances.back());
-        coordinates = TrimCoordinatesToLength(std::move(coordinates), offset, segment_distances);
+        coordinates =
+            TrimCoordinatesToLength(std::move(coordinates),
+                                    2 * distance_to_skip_over_due_to_coordinate_inaccuracies,
```
Comment (TheMarex) [inline src/extractor/guidance/coordinate_extractor.cpp]: What about `distance_epsilon`, this name maybe a little bit long, heh.

File: src/extractor/guidance/coordinate_extractor.cpp [hunk 5]
```diff
@@ -709,16 +715,17 @@ bool CoordinateExtractor::IsCurve(const std::vector<util::Coordinate> &coordinat
             return std::make_tuple(
                 true, 1, get_deviation(coordinates.front(), coordinates.back(), coordinates[1]));
 
-        const auto maximum_itr =
+        const auto one_past_maximum_itr =
```
Comment (TheMarex) [inline src/extractor/guidance/coordinate_extractor.cpp]: `itr` -> `iter`

File: src/extractor/guidance/coordinate_extractor.cpp (hunk 4 above)
Comment (danpat) [inline src/extractor/guidance/coordinate_extractor.cpp]: I dunno, I actually think this is helpful naming - in 6 months from now, when someone else is looking at this code, the intent is pretty clear.

Makes it easy to search for too :-)

File: src/extractor/guidance/coordinate_extractor.cpp [hunk 6]
```diff
@@ -337,6 +337,24 @@ util::Coordinate CoordinateExtractor::ExtractRepresentativeCoordinate(
         return result;
     }
 
+    if (IsDirectOffset(coordinates,
```
Comment (danpat) [inline src/extractor/guidance/coordinate_extractor.cpp]: Would you be able to add a comment that describes why `IsDirectOffset` needs to happen before `IsCurve`?  That's not obvious to me, and it seems like an important part of this change.

File: src/extractor/guidance/coordinate_extractor.cpp (hunk 4 above)
Comment (chaupow) [inline src/extractor/guidance/coordinate_extractor.cpp]: :D or name it `distance_epsilon` and put the explanation in a comment.

gotta agree with danpat tho. its intent is very clear.

**Pull request merged**